File: ai/execution/queryExecutor.js
import { normalizeTimeRange } from './timeRangeNormalizer.js';

const DEFAULT_TRANSACTION_LIST_LIMIT = 10;
const DEFAULT_FIRST_N_LIMIT = 1;
const MAX_TRANSACTION_LIST_LIMIT = 100;

const normalizeListLimit = (value, fallback) => {
  if (!Number.isInteger(value) || value <= 0) return fallback;
  return Math.min(value, MAX_TRANSACTION_LIST_LIMIT);
};

export class QueryExecutor {
  constructor({ transactionRepository, accountService, profileService } = {}) {
    this.transactionRepository = transactionRepository;
    this.accountService = accountService;
    this.profileService = profileService;
  }

  async execute({ userId, userEmail = null, query }) {
    if (!query?.domain || !query?.intent) {
      throw new Error('Invalid structured query payload');
    }

    if (query.domain === 'transactions' && query.intent === 'transactions_query') {
      return this.executeTransactionsQuery({ userId, userEmail, query });
    }

    if (query.domain === 'account' && query.intent === 'get_balance') {
      const result = await this.accountService.getBalance({ userId });
      return { operation: 'get_balance', result };
    }

    if (query.domain === 'profile' && query.intent === 'get_user_name') {
      const result = await this.profileService.getIdentity({ userId });
      return { operation: 'get_user_identity', result };
    }

    throw new Error(`Unsupported domain/intent: ${query.domain}/${query.intent}`);
  }

  async executeTransactionsQuery({ userId, userEmail = null, query }) {
    if (!this.transactionRepository) {
      throw new Error('transactionRepository is required for transactions_query');
    }

    let normalizedRange;
    try {
      normalizedRange = normalizeTimeRange({
        dateRange: query.dateRange
      });
    } catch {
      return {
        operation: 'get_recent_transfers',
        result: { found: false, message: 'Invalid date range' }
      };
    }

    const { startDate, endDate } = normalizedRange;
    const baseArgs = {
      userId,
      userEmail,
      filters: query.filters || {},
      startDate,
      endDate
    };

    if (query.aggregation === 'counterparty') {
      const recipientName = String(query.recipientName || '').trim();
      if (!recipientName) {
        return {
          operation: 'get_last_sent_transfer_to_recipient',
          result: { found: false, message: 'recipientName is required' }
        };
      }

      const limit = Number.isInteger(query.limit) && query.limit > 0 ? query.limit : 10;
      const items = await this.transactionRepository.listCounterpartyByName({
        userId,
        userEmail,
        recipientName,
        limit,
        startDate,
        endDate
      });
      return {
        operation: 'get_last_sent_transfer_to_recipient',
        result: { found: true, recipientName, count: items.length, items }
      };
    }

    if (query.aggregation === 'count') {
      const count = await this.transactionRepository.countBySemanticQuery(baseArgs);
      return {
        operation: 'count_transfers',
        result: { found: true, count, from: startDate, to: endDate }
      };
    }

    if (query.aggregation === 'first_n') {
      const limit = normalizeListLimit(query.limit, DEFAULT_FIRST_N_LIMIT);
      const items = await this.transactionRepository.listBySemanticQuery({ ...baseArgs, limit, sort: 'desc' });
      return {
        operation: 'get_first_n_transfers',
        result: { found: true, count: items.length, items, from: startDate, to: endDate }
      };
    }

    if (query.aggregation === 'list' || !query.aggregation) {
      const limit = normalizeListLimit(query.limit, DEFAULT_TRANSACTION_LIST_LIMIT);
      const items = await this.transactionRepository.listBySemanticQuery({ ...baseArgs, limit, sort: 'desc' });
      return {
        operation: 'get_recent_transfers',
        result: { found: true, count: items.length, items, from: startDate, to: endDate }
      };
    }

    throw new Error(`Unsupported aggregation: ${query.aggregation}`);
  }
}

File: ai/services/query_executor.py
from ai.services.time_range_normalizer import normalize_time_range

DEFAULT_TRANSACTION_LIST_LIMIT = 10
DEFAULT_FIRST_N_LIMIT = 1
MAX_TRANSACTION_LIST_LIMIT = 100


def _normalize_list_limit(value, fallback: int) -> int:
    if not isinstance(value, int) or value <= 0:
        return fallback
    return min(value, MAX_TRANSACTION_LIST_LIMIT)


def _normalize_sort_direction(value, fallback="desc") -> str:
    return value if value in ("asc", "desc") else fallback


def _build_result_meta(*, start_date=None, end_date=None, limit=None, sort_direction=None) -> dict:
    from_iso = start_date.isoformat() if start_date else None
    to_iso = end_date.isoformat() if end_date else None
    return {
        "from": from_iso,
        "to": to_iso,
        "hasDateRange": bool(start_date or end_date),
        "requestedLimit": limit,
        "sortDirection": sort_direction,
    }


class QueryExecutor:
    def __init__(self, *, transaction_repository=None, account_service=None, profile_service=None):
        self.transaction_repository = transaction_repository
        self.account_service = account_service
        self.profile_service = profile_service

    async def execute(self, *, user_id, user_email=None, query: dict) -> dict:
        if not query or not query.get("domain") or not query.get("intent"):
            raise ValueError("Invalid structured query payload")

        if query["domain"] == "transactions" and query["intent"] == "transactions_query":
            return await self._execute_transactions_query(user_id=user_id, user_email=user_email, query=query)

        if query["domain"] == "account" and query["intent"] == "get_balance":
            result = await self.account_service.get_balance(user_id=user_id)
            return {"operation": "get_balance", "result": result}

        if query["domain"] == "profile" and query["intent"] == "get_user_name":
            result = await self.profile_service.get_identity(user_id=user_id)
            return {"operation": "get_user_identity", "result": result}

        raise ValueError(f"Unsupported domain/intent: {query['domain']}/{query['intent']}")

    async def _execute_transactions_query(self, *, user_id, user_email=None, query: dict) -> dict:
        if not self.transaction_repository:
            raise ValueError("transaction_repository is required for transactions_query")

        try:
            normalized_range = normalize_time_range(date_range=query.get("dateRange"))
        except Exception:
            return {"operation": "get_recent_transfers", "result": {"found": False, "message": "Invalid date range"}}

        start_date = normalized_range["startDate"]
        end_date = normalized_range["endDate"]
        base_args = dict(
            user_id=user_id,
            user_email=user_email,
            filters=query.get("filters") or {},
            start_date=start_date,
            end_date=end_date,
        )

        aggregation = query.get("aggregation")

        if aggregation == "counterparty":
            recipient_name = str(query.get("recipientName") or "").strip()
            if not recipient_name:
                return {"operation": "get_last_sent_transfer_to_recipient", "result": {"found": False, "message": "recipientName is required"}}
            limit = query.get("limit") if isinstance(query.get("limit"), int) and query["limit"] > 0 else 10
            items = self.transaction_repository.list_counterparty_by_name(
                user_id=user_id, user_email=user_email, recipient_name=recipient_name,
                limit=limit, start_date=start_date, end_date=end_date
            )
            return {
                "operation": "get_last_sent_transfer_to_recipient",
                "result": {
                    "found": True,
                    "recipientName": recipient_name,
                    "count": len(items),
                    "items": items,
                    **_build_result_meta(start_date=start_date, end_date=end_date, limit=limit, sort_direction="desc"),
                },
            }

        if aggregation == "count":
            count = self.transaction_repository.count_by_semantic_query(**base_args)
            return {
                "operation": "count_transfers",
                "result": {"found": True, "count": count, **_build_result_meta(start_date=start_date, end_date=end_date)},
            }

        if aggregation == "first_n":
            limit = _normalize_list_limit(query.get("limit"), DEFAULT_FIRST_N_LIMIT)
            sort = _normalize_sort_direction(query.get("sortDirection"), "desc")
            items = self.transaction_repository.list_by_semantic_query(**base_args, limit=limit, sort=sort)
            return {
                "operation": "get_first_n_transfers",
                "result": {
                    "found": True,
                    "count": len(items),
                    "items": items,
                    **_build_result_meta(start_date=start_date, end_date=end_date, limit=limit, sort_direction=sort),
                },
            }

        if aggregation in ("list", None, ""):
            limit = _normalize_list_limit(query.get("limit"), DEFAULT_TRANSACTION_LIST_LIMIT)
            sort = _normalize_sort_direction(query.get("sortDirection"), "desc")
            items = self.transaction_repository.list_by_semantic_query(**base_args, limit=limit, sort=sort)
            return {
                "operation": "get_recent_transfers",
                "result": {
                    "found": True,
                    "count": len(items),
                    "items": items,
                    **_build_result_meta(start_date=start_date, end_date=end_date, limit=limit, sort_direction=sort),
                },
            }

        raise ValueError(f"Unsupported aggregation: {aggregation}")
